pkg/plugin/plugin.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"path"
+	"strings"
+
+	"github.com/spf13/pflag"
+)
+
+type Base interface {
+	// Name returns a name defining the plugin type.
+	// For example, Kubebuilder's plugins would return "go".
+	Name() string
+	// Version returns the project version that this plugin implements.
+	// For example, Kubebuilder's Go v2 plugin implementation would return 2.
+	Version() string
+}
+
+func Key(p Base) string {
+	return path.Join(p.Name(), "v"+strings.TrimLeft(p.Version(), "v"))
+}
+
+type InjectsCommandName interface {
```

Comment:
can we just switch this to the context-based approach?

Reply:
Done.